frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx

```diff
@@ -271,41 +323,33 @@ function DataFrame({
    * @param triggerRerun - Whether to trigger a rerun of the script after applying edits
    */
   const applyEdits = React.useCallback(
-    (clearSelections = false, triggerRerun = true) => {
+    // Use debounce to prevent rapid updates to the widget state.
+    debounce(DEBOUNCE_TIME_MS, () => {
       if (numRows !== editingState.current.getNumRows()) {
         // Reset the number of rows if it has been changed in the editing state
         setNumRows(editingState.current.getNumRows())
       }
 
-      if (clearSelections) {
-        clearSelection()
+      const currentEditingState = editingState.current.toJson(columns)
+      let currentWidgetState = widgetMgr.getStringValue(element as WidgetInfo)
+
+      if (currentWidgetState === undefined) {
+        // Create an empty widget state
+        currentWidgetState = new EditingState(0).toJson([])
       }
 
-      // Use debounce to prevent rapid updates to the widget state.
-      debounce(DEBOUNCE_TIME_MS, () => {
-        const currentEditingState = editingState.current.toJson(columns)
-        let currentWidgetState = widgetMgr.getStringValue(
-          element as WidgetInfo
+      // Only update if there is actually a difference between editing and widget state
+      if (currentEditingState !== currentWidgetState) {
+        widgetMgr.setStringValue(
+          element as WidgetInfo,
+          currentEditingState,
+          {
+            fromUi: true,
+          },
+          fragmentId
```

Comment:
This block seems to exist a couple of times, e.g. [here](https://github.com/streamlit/streamlit/pull/8411/files#diff-eda1f0faafab9bbafe2c7b8b34241c35192488443ab3e02f0291d91e40ef632fR269-R282)

Maybe it'd make sense to move it to an extra function

```typescript
function setDataFrameWidgetState(widgetMgr, oldState, newState, fragmentId) {
if (newState === undefined || oldState === newState) {
  return
}
        widgetMgr.setStringValue(
          element as WidgetInfo,
          currentEditingState,
          {
            fromUi: true,
          },
          fragmentId);
}
```
and call this function from the different places.